python_modules/dagster/dagster/_core/execution/calculate_data_time.py

```diff
@@ -0,0 +1,125 @@
+            if not upstream_key_strs:
+                return {key_str}
+            return set().union(
+                *(_recursive(upstream_key_str) for upstream_key_str in upstream_key_strs)
+            )
+
+        return _recursive(asset_key.to_user_string())
+
+    def _get_upstream_materialization_times(record):
+
+        # check if we've already calculated data times for this record
+        cur_tags = instance.get_asset_event_tags(record.storage_id)
+        if ROOT_DATA_TAG in cur_tags:
+            return cur_tags[ROOT_DATA_TAG]
+
+        # find parent asset keys
+        cur_key = record.event_log_entry.dagster_event.event_specific_data.materialization.asset_key
+        upstream_keys = upstream_asset_key_mapping[cur_key.to_user_string()]
+
+        if not upstream_keys:
+            # you're a root key, so no need to recurse
+            root_data = {
+                cur_key.to_user_string(): (
+                    record.storage_id,
+                    record.event_log_entry.timestamp,
+                )
+            }
+        else:
+            root_data = {}
+            for upstream_key_str in upstream_keys:
+                upstream_key = AssetKey.from_user_string(upstream_key_str)
+                upstream_records = instance.get_event_records(
+                    EventRecordsFilter(
+                        event_type=DagsterEventType.ASSET_MATERIALIZATION,
+                        asset_key=upstream_key,
+                        before_cursor=record.storage_id,
+                    ),
+                    ascending=False,
+                    limit=1,
+                )
+                if not upstream_records:
+                    # no materialization happened before this one
+                    # in these cases, set the root data times to None as no upstream data was used
+                    # for this materialization (as far as we can tell)
+                    for root_key in _get_root_keys(upstream_key):
+                        root_data[root_key] = (None, None)
+
+                else:
+                    # recurse to find the materialization times used to generated your parents
+                    upstream_times = _get_upstream_materialization_times(upstream_records[0])
+                    for key, tup in upstream_times.items():
+                        tup = tuple(tup)
+                        # if root data is missing, override other values
+                        if tup == (None, None) or root_data.get(key) == (None, None):
+                            root_data[key] = (None, None)
+                        else:
+                            root_data[key] = max(root_data.get(key, tup), tup)
+
+        # now that we've done this calculation, store it in the database
+        instance.add_asset_event_tags(record.storage_id, {ROOT_DATA_TAG: root_data})
+        return root_data
+
+    # get most recent asset materialization event record for this asset
+    records = instance.get_event_records(
```

Comment:
I think you can might be able to use `instance.get_latest_materialization_event`?

Reply:
I wish :( -- we need the storage ID, and get_latest_materialization_event only returns the EventLogEntry, not the EventLogRecord (which contains the storage id)